File: stdlib/lua/hull/middleware/tests/test_logger.lua
-- test_logger.lua — Tests for hull.middleware.logger
--
-- Tests pure-function helpers (no runtime globals needed).
-- Run via: the C test harness (test_lua_runtime.c) loads and executes this.

local logger = require('hull.middleware.logger')

local pass = 0
local fail = 0

local function test(name, fn)
    local ok, err = pcall(fn)
    if ok then
        pass = pass + 1
    else
        fail = fail + 1
        print("FAIL: " .. name .. ": " .. tostring(err))
    end
end

local function assert_eq(a, b, msg)
    if a ~= b then
        error((msg or "") .. " expected " .. tostring(b) .. ", got " .. tostring(a))
    end
end

-- ── generate_id ──────────────────────────────────────────────────────

test("generate_id returns hex string", function()
    local id = logger.generate_id()
    assert(type(id) == "string", "expected string")
    assert(#id > 0, "expected non-empty")
end)

test("generate_id increments", function()
    local a = logger.generate_id()
    local b = logger.generate_id()
    assert(a ~= b, "expected different IDs")
end)

-- ── format_line ──────────────────────────────────────────────────────

test("format_line basic pairs", function()
    local line = logger.format_line({
        { "method", "GET" },
        { "path", "/api" },
    })
    assert_eq(line, "method=GET path=/api")
end)

test("format_line quotes values with spaces", function()
    local line = logger.format_line({
        { "ua", "Mozilla Firefox" },
    })
    assert_eq(line, 'ua="Mozilla Firefox"')
end)

-- ── should_skip ──────────────────────────────────────────────────────

test("should_skip matches exact path", function()
    assert_eq(logger.should_skip("/health", {"/health"}), true)
end)

test("should_skip returns false on miss", function()
    assert_eq(logger.should_skip("/api", {"/health"}), false)
end)

test("should_skip returns false for nil list", function()
    assert_eq(logger.should_skip("/health", nil), false)
end)

-- ── middleware factory ───────────────────────────────────────────────

test("middleware returns a function", function()
    local mw = logger.middleware({})
    assert(type(mw) == "function", "expected function")
end)

return {pass = pass, fail = fail}

File: stdlib/js/hull/middleware/tests/test_logger.js
// test_logger.js — Tests for hull:middleware:logger
//
// Tests pure-function helpers (no runtime globals needed).

import { logger } from "hull:middleware:logger";

let pass = 0;
let fail = 0;

function test(name, fn) {
    try {
        fn();
        pass++;
    } catch (e) {
        fail++;
        print("FAIL: " + name + ": " + e.message);
    }
}

function assertEq(a, b, msg) {
    if (a !== b)
        throw new Error((msg || "") + " expected " + b + ", got " + a);
}

// ── generateId ───────────────────────────────────────────────────────

test("generateId returns hex string", () => {
    const id = logger.generateId();
    assertEq(typeof id, "string");
    if (id.length === 0) throw new Error("expected non-empty");
});

test("generateId increments", () => {
    const a = logger.generateId();
    const b = logger.generateId();
    if (a === b) throw new Error("expected different IDs");
});

// ── formatLine ───────────────────────────────────────────────────────

test("formatLine basic pairs", () => {
    const line = logger.formatLine([
        ["method", "GET"],
        ["path", "/api"],
    ]);
    assertEq(line, "method=GET path=/api");
});

test("formatLine quotes values with spaces", () => {
    const line = logger.formatLine([
        ["ua", "Mozilla Firefox"],
    ]);
    assertEq(line, 'ua="Mozilla Firefox"');
});

// ── shouldSkip ───────────────────────────────────────────────────────

test("shouldSkip matches exact path", () => {
    assertEq(logger.shouldSkip("/health", ["/health"]), true);
});

test("shouldSkip returns false on miss", () => {
    assertEq(logger.shouldSkip("/api", ["/health"]), false);
});

test("shouldSkip returns false for null list", () => {
    assertEq(logger.shouldSkip("/health", null), false);
});

// ── middleware factory ───────────────────────────────────────────────

test("middleware returns a function", () => {
    const mw = logger.middleware({});
    assertEq(typeof mw, "function");
});

export default { pass, fail };
